### ISSUE TITLE
Link split using pre-define edge indices

### ISSUE THREAD

**Issue by shusha1212** [pyg-team/pytorch_geometric#6684]
### 📚 Describe the documentation issue

Hi there,
I'm using pyg for a link prediction problem in heterograph and used the official Doc. as well as your [post ](https://medium.com/@pytorch_geometric/link-prediction-on-heterogeneous-graphs-with-pyg-6d5c29677c70) as a code reference.
I need to split my data (edges) into train and test sets. 
Following your documentation, I have implemented the toy example below but encountered several gaps in the doc./code.

**limitations/clarifications needed:**
-
1. How to perform a link split of the entire data using pre-defined edge indices for train and test (instead of random edge selection)?
2. How to include known negative labels in the test data (instead of random negative sampling)?
3. How to include known negative labels in the train data?
        
#      Toy Example
        edges = {
            'train': {
                'label': [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
                'label_index_0': [44, 49, 50, 51, 20, 44, 44, 20, 66, 20, 20, 12],
                'label_index_1': [1039, 2638, 2639, 2640, 663, 1370, 1435, 718, 2226, 1353, 1354, 598]
            },
            'test': {
                'label': [1, 1, 1, 1, 0, 0, 0, 0],
                'label_index_0': [42, 66, 3, 3, 66, 66, 44, 44],
                'label_index_1': [209, 2163, 2722, 3919, 393, 566, 992, 993],
                'pos_label_index_0': [42, 66, 3, 3],
                'pos_label_index_1': [209, 2163, 2722, 3919]
            },
        }

        #  ---------    split randomly   ---------
        data = HeteroData()
        #  Nodes
        data[cls.srna].node_id = torch.arange(len(cls.srna_nodes))
        data[cls.mrna].node_id = torch.arange(len(cls.mrna_nodes))
        #  Edges
        data_pos_edge_index = torch.stack([torch.from_numpy(np.array(edges['train']['label_index_0'] +
                                                                 edges['test']['pos_label_index_0'])),
                                       torch.from_numpy(np.array(edges['train']['label_index_1'] +
                                                                 edges['test']['pos_label_index_1']))], dim=0)
        data[cls.srna, cls.srna_to_mrna, cls.mrna].edge_index = data_pos_edge_index
        data = T.ToUndirected()(data)

        transform = RandomLinkSplit(
            num_val=0,
            num_test=0.25,
            disjoint_train_ratio=0.3,
            neg_sampling_ratio=1,
            add_negative_train_samples=False,
            edge_types=(cls.srna, cls.srna_to_mrna, cls.mrna),
            rev_edge_types=(cls.mrna, f"rev_{cls.srna_to_mrna}", cls.srna),
        )
        train_data, val_data, test_data = transform(data)
        logger.debug(f"\n All positive data:\n ============== \n{data}\n"
                     f"\n Train data:\n ============== \n{train_data}\n"
                     f"\n Test data:\n ============== \n{test_data}\n")

 **All positive data:**
HeteroData(
  srna={ node_id=[94] },
  mrna={ node_id=[4300] },
  (srna, targets, mrna)={ edge_index=[2, 16] },
  (mrna, rev_targets, srna)={ edge_index=[2, 16] }
)

 **Train data:**
HeteroData(
  srna={ node_id=[94] },
  mrna={ node_id=[4300] },
  (srna, targets, mrna)={
    edge_index=[2, 9],
    edge_label=[3],
    edge_label_index=[2, 3]
  },
  (mrna, rev_targets, srna)={ edge_index=[2, 9] }
)

 **Test data:**
HeteroData(
  srna={ node_id=[94] },
  mrna={ node_id=[4300] },
  (srna, targets, mrna)={
    edge_index=[2, 12],
    edge_label=[8],
    edge_label_index=[2, 8]
  },
  (mrna, rev_targets, srna)={ edge_index=[2, 12] }
)

 **My workaround**
-----
In order to be able to split the data using pre-defined edge indices for train and test sets, and also to be able to include
known negative labels in the test, I have implemented the workaround below.

        #  ---------    split by edge indices    ---------
        train_edge_label = torch.from_numpy(np.array(edges['train']['label']))
        train_edge_label_index = torch.stack([torch.from_numpy(np.array(edges['train']['label_index_0'])),
                                              torch.from_numpy(np.array(edges['train']['label_index_1']))], dim=0)
        test_edge_label = torch.from_numpy(np.array(edges['test']['label']))
        test_edge_label_index = torch.stack([torch.from_numpy(np.array(edges['test']['label_index_0'])),
                                             torch.from_numpy(np.array(edges['test']['label_index_1']))], dim=0)
        # ------  My Train Data
        my_train_data = HeteroData()
        #  Nodes
        my_train_data[cls.srna].node_id = torch.arange(len(cls.srna_nodes))
        my_train_data[cls.mrna].node_id = torch.arange(len(cls.mrna_nodes))
        #  Edges
        my_train_data[cls.srna, cls.srna_to_mrna, cls.mrna].edge_index = train_edge_label_index
        my_train_data = T.ToUndirected()(my_train_data)
        transform = RandomLinkSplit(
            num_val=0,
            num_test=0,
            disjoint_train_ratio=0.3,
            neg_sampling_ratio=0,
            add_negative_train_samples=False,
            edge_types=(cls.srna, cls.srna_to_mrna, cls.mrna),
            rev_edge_types=(cls.mrna, f"rev_{cls.srna_to_mrna}", cls.srna),
        )
        my_train_data, _, _ = transform(my_train_data)

        # ------  My Test Data
        my_test_data = HeteroData()
        # Nodes
        my_test_data[cls.srna].node_id = torch.arange(len(cls.srna_nodes))
        my_test_data[cls.mrna].node_id = torch.arange(len(cls.mrna_nodes))
        #  Edges
        my_test_data[cls.srna, cls.srna_to_mrna, cls.mrna].edge_index = train_edge_label_index  # all train's edges
        my_test_data[cls.srna, cls.srna_to_mrna, cls.mrna].edge_label = test_edge_label
        my_test_data[cls.srna, cls.srna_to_mrna, cls.mrna].edge_label_index = test_edge_label_index
        my_test_data = T.ToUndirected()(my_test_data)

        logger.debug(f"\n MY Train data:\n ============== \n{my_train_data}\n"
                     f"\n MY Test data:\n ============== \n{my_test_data}\n")

 **My Train data:**
HeteroData(
  srna={ node_id=[94] },
  mrna={ node_id=[4300] },
  (srna, targets, mrna)={
    edge_index=[2, 9],
    edge_label=[3],
    edge_label_index=[2, 3]
  },
  (mrna, rev_targets, srna)={ edge_index=[2, 9] }
)

 **My Test data:**
HeteroData(
  srna={ node_id=[94] },
  mrna={ node_id=[4300] },
  (srna, targets, mrna)={
    edge_index=[2, 12],
    edge_label=[8],
    edge_label_index=[2, 8]
  },
  (mrna, rev_targets, srna)={ edge_index=[2, 12] }
)

 **Questions**
---

1. Is this workaround valid?
2. Is this the right way to include known negative labels in the test data?
3. open - how to include known negative labels in the train data?

**Comment by rusty1s** [comment]
Sorry for missing out on this issue. I think if you have a pre-defined edge split, it's the most natural way to do it by yourself. I am not really seeing how you could leverage `RandomLinkSplit` here. This sounds somewhat hacky.
```python
train_data = copy.copy(data)
train_data['srna', 'mrna'].edge_label = ...
train_data['srna', 'mrna'].edge_label_index = ...
```
Note that we ideally also need to remove any edges in `edge_index` that also appear in `edge_label_index`.

Let me know if I am missing something.

**Comment by shusha1212** [comment]
Hi @rusty1s,
Thank you so much for your reply 👍 
So are you basically suggesting the solution below?
Will very much appreciate your review 

##  Labels
```
edges = {
            'train': {
                'label': [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
                'label_index_0': [44, 49, 50, 51, 20, 44, 44, 20, 66, 20, 20, 12],
                'label_index_1': [1039, 2638, 2639, 2640, 663, 1370, 1435, 718, 2226, 1353, 1354, 598]
            },
            'test': {
                'label': [1, 1, 1, 1, 0, 0, 0, 0],
                'label_index_0': [42, 66, 3, 3, 66, 66, 44, 44],
                'label_index_1': [209, 2163, 2722, 3919, 393, 566, 992, 993],
                'pos_label_index_0': [42, 66, 3, 3],
                'pos_label_index_1': [209, 2163, 2722, 3919]
            },
        }
```
##  Train Data 
```
train_data = HeteroData()
# ------  Nodes
train_data[cls.srna].node_id = torch.arange(len(cls.srna_nodes))
train_data[cls.mrna].node_id = torch.arange(len(cls.mrna_nodes))

# ------  Edges
# ---  9/12 train edges for message passing
train_edge_index = torch.stack([torch.from_numpy(np.array(edges['train']['label_index_0'][:9])),
                               torch.from_numpy(np.array(edges['train']['label_index_1'][:9]))], dim=0)
train_data[cls.srna, cls.srna_to_mrna, cls.mrna].edge_index = train_edge_index
train_data = T.ToUndirected()(train_data)

# ---  3/12 train edges for supervision
train_data[cls.srna, cls.srna_to_mrna, cls.mrna].edge_label = torch.from_numpy(np.array(edges['train']['label'][9:]))
train_data[cls.srna, cls.srna_to_mrna, cls.mrna].edge_label_index = \
    torch.stack([torch.from_numpy(np.array(edges['train']['label_index_0'][9:])),
    torch.from_numpy(np.array(edges['train']['label_index_1'][9:]))], dim=0)
```
##  Test Data 
```
test_data = HeteroData()
# ------  Nodes
test_data[cls.srna].node_id = torch.arange(len(cls.srna_nodes))
test_data[cls.mrna].node_id = torch.arange(len(cls.mrna_nodes))

# ------  Edges
# ---  all train edges for message passing
test_edge_index = torch.stack([torch.from_numpy(np.array(edges['train']['label_index_0'])),
                               torch.from_numpy(np.array(edges['train']['label_index_1']))], dim=0)
test_data[cls.srna, cls.srna_to_mrna, cls.mrna].edge_index = test_edge_index
test_data = T.ToUndirected()(test_data)

# ---  test edges for supervision
test_data[cls.srna, cls.srna_to_mrna, cls.mrna].edge_label = torch.from_numpy(np.array(edges['test']['label']))
test_data[cls.srna, cls.srna_to_mrna, cls.mrna].edge_label_index = \
    torch.stack([torch.from_numpy(np.array(edges['test']['label_index_0'])),
    torch.from_numpy(np.array(edges['test']['label_index_1']))], dim=0)
`````

**Comment by rusty1s** [comment]
Yes, this looks perfect.
